best-practices-cel/require-drop-all/require-drop-all.yaml

```diff
@@ -0,0 +1,58 @@
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: drop-all-capabilities
+  annotations:
+    policies.kyverno.io/title: Drop All Capabilities in CEL expressions
+    policies.kyverno.io/category: Best Practices in CEL 
+    policies.kyverno.io/severity: medium
+    kyverno.io/kubernetes-version: "1.26-1.27"
+    policies.kyverno.io/minversion: 1.11.0
+    policies.kyverno.io/subject: Pod
+    policies.kyverno.io/description: >-
+      Capabilities permit privileged actions without giving full root access. All
+      capabilities should be dropped from a Pod, with only those required added back.
+      This policy ensures that all containers explicitly specify the `drop: ["ALL"]`
+      ability. Note that this policy also illustrates how to cover drop entries in any
+      case although this may not strictly conform to the Pod Security Standards.
+spec:
+  validationFailureAction: Audit
+  background: true
+  rules:
+    - name: require-drop-all
+      match:
+        any:
+        - resources:
+            kinds:
+              - Pod
+      celPreconditions:
+        - expression: "request.operation != 'DELETE'"
+          name: "Operation should not be DELETE"
+      validate:
+        cel:
+          expressions:
+            - expression: >-
+                object.spec.containers.all(container, 
+                has(container.securityContext) &&
+                has(container.securityContext.capabilities) &&
+                has(container.securityContext.capabilities.drop) &&
+                container.securityContext.capabilities.drop.exists(capability, capability.upperAscii() == 'ALL'))
```

Comment:
We could make use of `variables` since this expression is duplicated for each of containers, init containers and ephemeral containers.

```
validate:
  cel:
    variables:
    - name: capabilities
      expression: "securityContext.capabilities"
    - name: drop
      expression: "variables.capabilities.drop"
    - name: is-drop-exists
      expression: "variables.drop.exists(capability, capability.upperAscii() == 'ALL'))"
```

Then, they can be used as follows:
```
validate:
 cel:
   expressions:
   - expression: >-
        object.spec.containers.all(container, 
        has(container.securityContext) &&
        has(container.variables.capabilities) &&
        has(container.variables.drop) &&
        container.variables. is-drop-exists)
```

I am not sure if this will work tbh but let us try to use variables whenever possible to reduce duplications and to be easily read afterwards.

Reply:
When we try the above we get an error:
```
drop-all-capabilities:
          require-drop-all: 'expression ''object.spec.containers.all(container,  has(container.securityContext)
            && has(container.variables.capabilities) && has(container.variables.drop) && container.variables.isDropExists)''
            resulted in error: no such key: variables'
```
Actually, the variables being used would be dependent on whether it is in the context of a container, initContainer or an ephemeralContainer. A single variable cannot be used for all contexts in this case. I don't see a way to use variables here. Please let me know if you have any ideas.

Reply:
I have made some [changes](https://github.com/kyverno/policies/blob/cddee324b552123804c82d9d83bf979850fafd9d/best-practices-cel/require-drop-all/require-drop-all.yaml#L31-#L43) to the policy. The upside is that it avoids duplicate logic. The only downside I see is that we do not particularly know whether it is a container, initContainer, or ephemeralContainer we are referring to.